### PR TITLE
[IMP] procurement_purchase_no_grouping: Add company level policy

### PR DESCRIPTION
**Pull request by Rad0van**
This is my attempt to create a slightly enhanced version of procurement_purchase_no_grouping. The idea here is to provide also System default setting that will cause the logic fall back to system-wide setting set up in System settings. The reason for this is we have more than 250k products in more than 300 categories so setting up each category would be rather laborious. This should also work for existing users - they only need to setup the system-wide default.

### CODE REVIEW

**Comment by Rad0van** [comment]
@pedrobaeza, @sergio-teruel  and @carlosdauden could maybe have a look and leave some comments? Thank you.

File: procurement_purchase_no_grouping/__manifest__.py
Comment (pedrobaeza) [inline procurement_purchase_no_grouping/__manifest__.py]: Don't bump version yourself, but let the ocabot to do it for you on merge.

File: procurement_purchase_no_grouping/models/product_category.py
Comment (pedrobaeza) [inline procurement_purchase_no_grouping/models/product_category.py]: Please don't change this: not putting anything is for using the upper configuration. The criteria order should be: check the product, if no policy, check the category, and if not, go to system parameter.

**Review by pedrobaeza — CHANGES_REQUESTED** [review]

File: procurement_purchase_no_grouping/models/product_category.py
Comment (Rad0van) [reply procurement_purchase_no_grouping/models/product_category.py]: @pedrobaeza not sure I follow this one. My findings looking into database is that when you install the module it will fill in the default value to all product categories. You can then change it to None but that kind of defeats my use case (which might be specific I admit). So if None is used to indicate to go one level higher then I guess there should be no default value. Or am I missing something?

Comment (pedrobaeza) [reply procurement_purchase_no_grouping/models/product_category.py]: Yes, both things: I don't want any default value nor being required.

Comment (Rad0van) [reply procurement_purchase_no_grouping/models/product_category.py]: > Yes, both things: I don't want any default value nor being required.

OK, that makes sense. So None will fall back to system default. Will do right away.

**Comment by Rad0van** [comment]
@pedrobaeza can you please help me here? Tests in some other modules fail as if there was no system setting - looks like this usually:

`File "/home/travis/build/OCA/purchase-workflow/purchase_request/models/stock_rule.py", line 92, in _run_buy
return super(StockRule, self)._run_buy(procurements)
File "/home/travis/build/OCA/purchase-workflow/procurement_purchase_no_grouping/models/stock_rule.py", line 19, in _run_buy
)["procured_purchase_grouping"]
KeyError: 'procured_purchase_grouping'`

I am scratching my head here - not understanding why this works in my module... Any help would be appreciated. Thank you.

**Comment by HviorForgeFlow** [comment]
I think the problem is that product categories that are loaded before you install the module do not fetch default value correctly.
Maybe you need some init hook to initialize that information... not sure, maybe @pedrobaeza can confirm this issue.
@Rad0van 

I would store that config on company level, but maybe there is another way to do it...

**Comment by Rad0van** [comment]
> I think the problem is that product categories that are loaded before you install the module do not fetch default value correctly.
> Maybe you need some init hook to initialize that information... not sure, maybe @pedrobaeza can confirm this issue.
> @Rad0van
> 
> I would store that config on company level, but maybe there is another way to do it...

@HviorForgeFlow hmm, but it seems to me it is failing on retrieving the system setting. Something like that happened once to me when I was testing - that initially the value wasn't there for some reason even it had default. But this happens randomly through the tests.

File: procurement_purchase_no_grouping/models/res_config_settings.py
Comment (pedrobaeza) [inline procurement_purchase_no_grouping/models/res_config_settings.py]: You should store this other way, as this means that new categories will take this value instead of empty one, which can't be the desired behavior, as it's not the same having "system wide grouping" than the last default that it was. I think a better granularity is to have this selection at res.company, at put here a related field.

Comment (Rad0van) [reply procurement_purchase_no_grouping/models/res_config_settings.py]: > You should store this other way, as this means that new categories will take this value instead of empty one, which can't be the desired behavior, as it's not the same having "system wide grouping" than the last default that it was. I think a better granularity is to have this selection at res.company, at put here a related field.

@pedrobaeza will (try to) do. But does this really solve the failed tests? In them it seems as if the system settings in some of the tests wasn't there at all.

Comment (pedrobaeza) [reply procurement_purchase_no_grouping/models/res_config_settings.py]: Well, other approach, other tests, so it's worth trying. Anyway, when you have it, I'll see specifically possible problems

**Comment by Rad0van** [comment]
@pedrobaeza well it seems to work now - you were right ;-) Now, should I also rework it as you suggested - i.e. put the setting into res.company?  So that it would be specific per res.company? Looking at my installations there aren't that many such things at res.company levels. Can you please be more specific? Will gladly do in  away you suggest.

**Comment by pedrobaeza** [comment]
In a multi-company environment, it can make sense to have a different grouping policy per company, and thus, the value is stored at res.company level instead of a system wide level (ir.config_parameter). There are several fields in res.company in core that follows this way.

File: procurement_purchase_no_grouping/models/res_config_settings.py
Comment (bealdav) [inline procurement_purchase_no_grouping/models/res_config_settings.py]: no need Class name and self if it's the same than current class in python3

Comment (Rad0van) [reply procurement_purchase_no_grouping/models/res_config_settings.py]: Good point for a Python newbie ;-) Thanx. Will do.

**Comment by Rad0van** [comment]
> In a multi-company environment, it can make sense to have a different grouping policy per company, and thus, the value is stored at res.company level instead of a system wide level (ir.config_parameter). There are several fields in res.company in core that follows this way.

@pedrobaeza will look into that. Now I squashed all the commits to one.

**Comment by Rad0van** [comment]
@pedrobaeza done. I hope at least ;-)

File: procurement_purchase_no_grouping/models/res_config_settings.py
Comment (pedrobaeza) [inline procurement_purchase_no_grouping/models/res_config_settings.py]: Don't add the help nor the string, as it's taken from the related field.

File: procurement_purchase_no_grouping/models/res_company.py
Comment (pedrobaeza) [inline procurement_purchase_no_grouping/models/res_company.py]: It's better to avoid this `required`, as in tests, this will be a nuisance, as this constraint is put on DB, but default is not applying until next run, so it provokes failures. Instead, put `required="1"` in the res_config_settings view for UI.

File: procurement_purchase_no_grouping/tests/test_procurement_purchase_no_grouping.py
Comment (pedrobaeza) [inline procurement_purchase_no_grouping/tests/test_procurement_purchase_no_grouping.py:90]: Why 2 times?

File: procurement_purchase_no_grouping/models/res_config_settings.py
Comment (Rad0van) [reply procurement_purchase_no_grouping/models/res_config_settings.py]: OK, will do.

File: procurement_purchase_no_grouping/models/res_company.py
Comment (Rad0van) [reply procurement_purchase_no_grouping/models/res_company.py]: But that would mean that unless someone sets it, it will be None? In combination with that Product Category does not have a default value we can run into situation where there will be no value at all - i.e. directly after installation.

File: procurement_purchase_no_grouping/tests/test_procurement_purchase_no_grouping.py
Comment (Rad0van) [reply procurement_purchase_no_grouping/tests/test_procurement_purchase_no_grouping.py:90]: Honestly - I wouldn't know. That's the way it is in original code. Was wondering when debugging why but let it go. Should that be limited to one?

File: procurement_purchase_no_grouping/models/res_company.py
Comment (pedrobaeza) [reply procurement_purchase_no_grouping/models/res_company.py]: No, required != default. The default will still be there. This is only a technical question for not setting the NOT NULL constraint on the DB, so you don't have problems in tests.

File: procurement_purchase_no_grouping/tests/test_procurement_purchase_no_grouping.py
Comment (pedrobaeza) [reply procurement_purchase_no_grouping/tests/test_procurement_purchase_no_grouping.py:90]: Yes, try to put only one and see if tests still pass

File: procurement_purchase_no_grouping/models/res_company.py
Comment (Rad0van) [reply procurement_purchase_no_grouping/models/res_company.py]: Still learning new thing. Thanx a lot. Will do.

File: procurement_purchase_no_grouping/tests/test_procurement_purchase_no_grouping.py
Comment (Rad0van) [reply procurement_purchase_no_grouping/tests/test_procurement_purchase_no_grouping.py:90]: OK.

Comment (Rad0van) [reply procurement_purchase_no_grouping/tests/test_procurement_purchase_no_grouping.py:90]: So it seems the double run was required.

**Comment by Rad0van** [comment]
@pedrobaeza All done except for the double procurement run. Without it even the original tests that didn't rely on default value failed.

**Review by pedrobaeza — APPROVED** [review]
OK, it's curious the double run thing.

Please change the commit message to:
```
[IMP] procurement_purchase_no_grouping: Add company level policy
```

**Comment by Rad0van** [comment]
@pedrobaeza thank you far valuable input.

**Comment by Rad0van** [comment]
@bealdav can you please re-review? Or maybe @sergio-teruel or @carlosdauden could have a look? Thank you.

**Comment by bealdav** [comment]
Yes I have to, but before I need to know more on the module that you amended, I don't use it up to now.

**Comment by Rad0van** [comment]
> Yes I have to, but before I need to know more on the module that you amended, I don't use it up to now.

Oh I thought you knew it. Before you could set that grouping type per Product Category I just added company level policy for Product Categories where no value is set.

**Comment by pedrobaeza** [comment]
Bad timing for this. My colleagues are on holidays. When they return, they can eventually review the PR.

**Comment by Rad0van** [comment]
> Bad timing for this. My colleagues are on holidays. When they return, they can eventually review the PR.

:-) Sure, no problem.

**Comment by HviorForgeFlow** [comment]
@Rad0van @pedrobaeza @sergio-teruel @carlosdauden any news on this?

**Comment by Rad0van** [comment]
> @Rad0van @pedrobaeza @sergio-teruel @carlosdauden any news on this?

Waiting for approval from others. Not much I can do now.

**Review by bealdav — APPROVED** [review]

**Review by sergio-teruel — APPROVED** [review]

**Pull request merged**
